File: PersianSwear.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class PersianSwear {
    private final ArrayList<String> swearWords = new ArrayList<>();

     public PersianSwear() {
         try (BufferedReader reader = new BufferedReader(new FileReader("data.txt"))) {
             String line;
             while ((line = reader.readLine()) != null) {
                 String[] splitedWords = line.split(", ");
                 swearWords.addAll(List.of(splitedWords));
             }
         } catch (IOException e) {
             throw new RuntimeException(e);
         }
    }

    private String clearWord(String word) {
        Pattern pattern = Pattern.compile("[\\p{P}\\p{Mn}\\p{S}\\p{javaDigit}]");
        Matcher matcher = pattern.matcher(word);
        matcher.reset();

        return matcher.replaceAll("").trim()
                .replace("\u200c", " ")
                .replace("ي", "ی")
                .replace("ك", "ک")
                .replace("ـ", "");
    }

    public void addWord(String word) {
        swearWords.add(word.trim());
    }

    public void addWords(String[] words) {
        swearWords.addAll(List.of(words));
    }

    public void removeWord(String word) {
        swearWords.remove(word.trim());
    }

    public void removeWords(String[] words) {
        swearWords.removeAll(List.of(words));
    }

    public boolean isBad(String word) {
        return swearWords.contains(clearWord(word));
    }

    public boolean hasSwear(String text) {
        List<String> splitedText = List.of(text.split(" "));

         for (String word : splitedText)
             if (swearWords.contains(clearWord(word)))
                 return true;

        return false;
    }

    public String filterWords(String text, String symbol) {
        List<String> splitedText = new ArrayList<>(List.of(text.split(" ")));

        for (String word : splitedText)
            if (swearWords.contains(clearWord(word)))
                splitedText.set(splitedText.indexOf(word), symbol);

        return String.join(" ", splitedText);
    }

    public String filterWords(String text) {
         return filterWords(text, "*");
    }
}

File: PersianSwear.go
package main

import (
	"bufio"
	"os"
	"regexp"
	"strings"
)

type PersianSwear struct {
	swearWords []string
}

func NewPersianSwear() *PersianSwear {
	ps := &PersianSwear{}
	file, err := os.Open("data.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		splitedWords := strings.Split(line, ", ")
		ps.swearWords = append(ps.swearWords, splitedWords...)
	}

	if err := scanner.Err(); err != nil {
		panic(err)
	}
	return ps
}

func (ps *PersianSwear) clearWord(word string) string {
	pattern := regexp.MustCompile(`[\p{P}\p{Mn}\p{S}\p{N}]`)
	clearedWord := pattern.ReplaceAllString(word, "")
	clearedWord = strings.TrimSpace(clearedWord)
	clearedWord = strings.ReplaceAll(clearedWord, "\u200c", " ")
	clearedWord = strings.ReplaceAll(clearedWord, "ي", "ی")
	clearedWord = strings.ReplaceAll(clearedWord, "ك", "ک")
	clearedWord = strings.ReplaceAll(clearedWord, "ـ", "")
	return clearedWord
}

func (ps *PersianSwear) AddWord(word string) {
	ps.swearWords = append(ps.swearWords, strings.TrimSpace(word))
}

func (ps *PersianSwear) AddWords(words ...string) {
	ps.swearWords = append(ps.swearWords, words...)
}

func (ps *PersianSwear) RemoveWord(word string) {
	trimmedWord := strings.TrimSpace(word)
	for i, w := range ps.swearWords {
		if w == trimmedWord {
			ps.swearWords = append(ps.swearWords[:i], ps.swearWords[i+1:]...)
			break
		}
	}
}

func (ps *PersianSwear) RemoveWords(words ...string) {
	for _, word := range words {
		ps.RemoveWord(word)
	}
}

func (ps *PersianSwear) IsBad(word string) bool {
	clearedWord := ps.clearWord(word)
	for _, w := range ps.swearWords {
		if w == clearedWord {
			return true
		}
	}
	return false
}

func (ps *PersianSwear) HasSwear(text string) bool {
	words := strings.Split(text, " ")
	for _, word := range words {
		if ps.IsBad(word) {
			return true
		}
	}
	return false
}

func (ps *PersianSwear) FilterWords(text, symbol string) string {
	words := strings.Split(text, " ")
	for i, word := range words {
		if ps.IsBad(word) {
			words[i] = symbol
		}
	}
	return strings.Join(words, " ")
}

func (ps *PersianSwear) FilterWordsWithDefaultSymbol(text string) string {
	return ps.FilterWords(text, "*")
}
